### PR TITLE
feat: support sourceLambda field for lambda-deployer to use existing Lambda

### PR DESCRIPTION
**Pull request by RonWang**
This commit addresses the feature requests from https://github.com/alexa/ask-cli/issues/166 and the partial feature request from https://github.com/alexa/ask-cli/issues/283.

### Summary
Introduce `userConfig.basedOn` into the skillInfrastructure JSON object in ask-resources.json. User can provide `basedOn.lambda.arn` to let ask-cli target the deployment to the existing Lambda ARN.

### Usage preview
```JSON
{
  "skillInfrastructure":{
    "userConfig":{
      "basedOn":{
        "lambda":{
          "arn":"arn:aws:lambda:us-east-1:12345:function:ask-function"
        }
      },
      "runtime":"nodejs10.x",
      "handler":"index.handler",
      "awsRegion":"us-east-1",
      "regionalOverrides":{
        "FE":{
          "basedOn":{
            "lambda":{
              "arn":"arn:aws:lambda:us-west-2:12345:function:ask-function"
            }
          }
        }
      }
    },
    "type":"@ask-cli/lambda-deployer"
  }
}
```

### Edge cases testing
1. When setting `basedOn.lambda.arn = "randomArn"` for default region, and FE region using regionalOverrides, the non-existing Lambda error will be displayed:
```
$ ask deploy
==================== Deploy Skill Metadata ====================
...
==================== Build Skill Code ====================
...
==================== Deploy Skill Infrastructure ====================
  ✖ Deploy Alexa skill infrastructure for region "default"
  ✔ Deploy Alexa skill infrastructure for region "EU"
  ✖ Deploy Alexa skill infrastructure for region "FE"
[Error]: CliError: Failed to load the Lambda (arn) specified in "basedOn". ResourceNotFoundException: Function not found: arn:aws:lambda:us-west-2:############:function:randomArn
Failed to load the Lambda (arn) specified in "basedOn". ResourceNotFoundException: Function not found: arn:aws:lambda:us-east-1:############:function:randomArn
```

2. When setting full Lambda ARN with a wrong AWS region (the example below set `us-east-1` LambdaARN to Alexa's region "FE"):
```
ask deploy
==================== Deploy Skill Metadata ====================
...
==================== Build Skill Code ====================
...
==================== Deploy Skill Infrastructure ====================
  ✔ Deploy Alexa skill infrastructure for region "default"
  ✔ Deploy Alexa skill infrastructure for region "EU"
  ✖ Deploy Alexa skill infrastructure for region "FE"
[Error]: CliError: Failed to load the Lambda (arn:aws:lambda:us-east-1:######:function:ask-existingLambda) specified in "basedOn". ResourceNotFoundException: Functions from 'us-east-1' are not reachable in this region ('us-west-2')
```

3. Reuse Lambda ARN but the trigger (which is linked to the skillId) doesn't match:
```
ask deploy
==================== Deploy Skill Metadata ====================
[Warn]: The hash of current skill package folder does not change compared to the last deploy hash result, CLI will skip the deploy of skill package.
Skill ID: amzn1.ask.skill.12345

==================== Build Skill Code ====================
...

==================== Deploy Skill Infrastructure ====================
  ✔ Deploy Alexa skill infrastructure for region "default"
  ✔ Deploy Alexa skill infrastructure for region "EU"
  ✔ Deploy Alexa skill infrastructure for region "FE"
[Error]: {
  "skill": {
    "resources": [
      {
        "action": "UPDATE",
        "errors": [
          {
            "message": "The trigger setting for the Lambda arn:aws:lambda:us-east-1:#####:function:ask-function is invalid."
          }
        ],
        "name": "Manifest",
        "status": "FAILED"
      }
    ],
    "skillId": "amzn1.ask.skill.12345"
  },
  "status": "FAILED"
}
```

### CODE REVIEW

File: lib/builtins/deploy-delegates/lambda-deployer/helper.js [hunk 1]
```diff
@@ -10,56 +10,86 @@ const retryUtils = require('@src/utils/retry-utility');
...
-function validateLambdaDeployState(reporter, awsProfile, awsRegion, currentRegionDeployState, ignoreHash, callback) {
-    const lambdaData = currentRegionDeployState.lambda;
-    if (R.isNil(lambdaData) || R.isEmpty(lambdaData) || R.isNil(lambdaData.arn)) {
-        return callback(null, { updatedDeployState: currentRegionDeployState });
+function loadLambdaInformation(reporter, loadLambdaConfig, callback) {
+    const { awsProfile, awsRegion, alexaRegion, deployState, userConfig, ignoreHash } = loadLambdaConfig;
+    const lambdaClient = new LambdaClient({ awsProfile, awsRegion });
+    // check if user sets userConfig with "sourceLambda" field to reuse existing Lambda
+    const sourceLambdaArn = alexaRegion === 'default' ? R.view(R.lensPath(['sourceLambda', 'arn']), userConfig)
+        : R.view(R.lensPath(['regionalOverrides', alexaRegion, 'sourceLambda', 'arn']), userConfig);
```
Comment (kakhaUrigashvili) [inline lib/builtins/deploy-delegates/lambda-deployer/helper.js:23]: optional(can be fixed during rewrite of lambda deployer)
can just use 
https://ramdajs.com/docs/#path

similar to other places.

File: lib/builtins/deploy-delegates/lambda-deployer/helper.js [hunk 2]
```diff
@@ -10,56 +10,86 @@ const retryUtils = require('@src/utils/retry-utility');
...
         if (err) {
             return callback(err);
         }
         // 1. check IAM role arn
-        const localIAMRole = currentRegionDeployState.iamRole;
+        const localIAMRole = deployState.iamRole;
         const remoteIAMRole = data.Configuration.Role;
         if (stringUtils.isNonBlankString(localIAMRole) && !R.equals(localIAMRole, remoteIAMRole)) {
-            return callback(`The IAM role for Lambda ARN (${lambdaArn}) should be ${remoteIAMRole}, but found ${localIAMRole}. \
+            return callback(`The IAM role for Lambda ARN (${lambdaArn}) should be "${remoteIAMRole}", but found "${localIAMRole}". \
```
Comment (kakhaUrigashvili) [inline lib/builtins/deploy-delegates/lambda-deployer/helper.js:46]: optional(can be fixed during rewrite of lambda deployer)
should be `new callback(new CliError(message))`?

same in other places where error is thrown.

**Review by kakhaUrigashvili — CHANGES_REQUESTED** [review]
@RonWang 

1) case 3 - Reuse Lambda ARN but the trigger (which is linked to the skillId) doesn't match.
Is it possible to fix it automatically during the deploy instead of throwing error?

2) please add example to the documentations about the sourceLambda

File: lib/builtins/deploy-delegates/lambda-deployer/helper.js (hunk 2 above)
Comment (RonWang) [reply lib/builtins/deploy-delegates/lambda-deployer/helper.js:46]: For this error class usage, fixing in the rewrite would be a better idea to me. I have added a TODO.

**Comment by kakhaUrigashvili** [comment]
> > 1. case 3 - Reuse Lambda ARN but the trigger (which is linked to the skillId) doesn't match.
> >    Is it possible to fix it automatically during the deploy instead of throwing error?
> > 2. please add example to the documentations about the sourceLambda
> 
> I will add the the doc in the next review. For the trigger, because it's about the event **permission**, I don't think we should modify it for an existing Lambda. An informative failure is best in my mind, since it might not be created by us, we cannot force its property to work with us. Does it sound good to you?

1) sounds good

2) just want to confirm that you will add docs before this PR is merged? I dont have any more concerns to this PR, just think we need few lines in the docs about this new feature.

**Review by kakhaUrigashvili — APPROVED** [review]

**Pull request merged**